engine/src/main/java/org/kigalisim/engine/SingleThreadEngine.java

```
@@ -1213,24 +1214,61 @@ private BigDecimal calculateAvailableRecycling(UseKey scope) {
+    }
+  }
+
+  /**
+   * Calculate recycling for a specific stage (EOL or RECHARGE).
+   *
+   * @param scope the scope to calculate recycling for
+   * @param availableUnits the units available for recycling at this stage
+   * @param stage the recovery stage (EOL or RECHARGE)
+   * @param unitConverter the unit converter to use
+   * @return the amount of recycling available in kg for this stage
+   */
+  private BigDecimal calculateRecyclingForStage(UseKey scope, BigDecimal availableUnits, RecoveryStage stage, UnitConverter unitConverter) {
+    try {
+      // Get stage-specific rates
+      EngineNumber recoveryRate = streamKeeper.getRecoveryRate(scope, stage);
+      EngineNumber yieldRate = streamKeeper.getYieldRate(scope, stage);
+
+      // Calculate rates as decimals
       BigDecimal recoveryRateDecimal = recoveryRate.getValue().divide(BigDecimal.valueOf(100));
       BigDecimal yieldRateDecimal = yieldRate.getValue().divide(BigDecimal.valueOf(100));
-      BigDecimal displacementRateDecimal = displacementRate.getValue().divide(BigDecimal.valueOf(100));
 
-      // Calculate recycling chain
-      BigDecimal retiredUnits = priorPopulation.getValue().multiply(retirementRateDecimal);
-      BigDecimal recoveredUnits = retiredUnits.multiply(recoveryRateDecimal);
+      // Calculate recycling chain for this stage
+      BigDecimal recoveredUnits = availableUnits.multiply(recoveryRateDecimal);
       BigDecimal recycledUnits = recoveredUnits.multiply(yieldRateDecimal);
 
       // Convert to kg
       EngineNumber initialCharge = streamKeeper.getInitialCharge(scope, "import");
       EngineNumber initialChargeKg = unitConverter.convert(initialCharge, "kg / unit");
       BigDecimal recycledKg = recycledUnits.multiply(initialChargeKg.getValue());
 
-      // Apply displacement rate
-      BigDecimal recyclingAvailable = recycledKg.multiply(displacementRateDecimal);
-
-      return recyclingAvailable;
+      return recycledKg;
     } catch (Exception e) {
```

Let's not use a catch all exceptions here with a default to zero.